### ISSUE TITLE
HDF5 on macOS: Cannot load 'redundant' attribute

### ISSUE THREAD

**Issue by jtramm** [openmc-dev/openmc#1072]
I've run into a bug when running on macOS that seems to be caused by HDF5. OpenMC compiles fine, but when I run I get the following error:

```
 Reading geometry XML file...
 Building neighboring cells lists for each surface...
 Reading U234 from /Users/jtramm/openmc/data/U234.h5
HDF5-DIAG: Error detected in HDF5 (1.10.2) thread 0:
  #000: H5A.c line 436 in H5Aopen(): unable to load attribute info from object header for attribute: 'redundant'
    major: Attribute
    minor: Unable to initialize object
  #001: H5Oattribute.c line 528 in H5O_attr_open_by_name(): can't locate attribute: 'redundant'
    major: Attribute
    minor: Object not found
HDF5-DIAG: Error detected in HDF5 (1.10.2) thread 0:
  #000: H5A.c line 681 in H5Aread(): not an attribute
    major: Invalid arguments to routine
    minor: Inappropriate type
HDF5-DIAG: Error detected in HDF5 (1.10.2) thread 0:
  #000: H5A.c line 1631 in H5Aclose(): not an attribute
    major: Invalid arguments to routine
    minor: Inappropriate type
HDF5-DIAG: Error detected in HDF5 (1.10.2) thread 0:
  #000: H5A.c line 436 in H5Aopen(): unable to load attribute info from object header for attribute: 'redundant'
    major: Attribute
    minor: Unable to initialize object
  #001: H5Oattribute.c line 528 in H5O_attr_open_by_name(): can't locate attribute: 'redundant'
    major: Attribute
    minor: Object not found
HDF5-DIAG: Error detected in HDF5 (1.10.2) thread 0:
  #000: H5A.c line 681 in H5Aread(): not an attribute
    major: Invalid arguments to routine
    minor: Inappropriate type
HDF5-DIAG: Error detected in HDF5 (1.10.2) thread 0:
  #000: H5A.c line 1631 in H5Aclose(): not an attribute
    major: Invalid arguments to routine
    minor: Inappropriate type
...
...
...
...
...
Reading c_H_in_H2O from /Users/jtramm/openmc/data/c_H_in_H2O.h5
 Maximum neutron transport energy: 2.00000E+07 eV for U235
 Reading tallies XML file...
 Writing summary.h5 file...
 Initializing source particles...

 ====================>     K EIGENVALUE SIMULATION     <====================

  Bat./Gen.      k       Entropy         Average k
  =========   ========   ========   ====================
At line 745 of file /Users/jtramm/openmc/src/physics.F90
Fortran runtime error: Index '2' of dimension 1 of array 'nuc%index_inelastic_scatter' above upper bound of 1
```

Upon inspection of reaction.cpp, I find the following lines seem to be the cause:

```C++
Reaction::Reaction(hid_t group, const std::vector<int>& temperatures)
{
  read_attribute(group, "Q_value", q_value_);
  read_attribute(group, "mt", mt_);
  int tmp;
  read_attribute(group, "center_of_mass", tmp);
  scatter_in_cm_ = (tmp == 1);
  read_attribute(group, "redundant", tmp);
  redundant_ = (tmp == 1);
  ...
```

When I look at my cross section hdf5 files, I am able to find the Q_value, mt, and center_of_mass attributes, but the "redundant" attribute is not present (at least on the nuclides I've browsed through). If I comment that out:

```C++
Reaction::Reaction(hid_t group, const std::vector<int>& temperatures)
{
  read_attribute(group, "Q_value", q_value_);
  read_attribute(group, "mt", mt_);
  int tmp;
  read_attribute(group, "center_of_mass", tmp);
  scatter_in_cm_ = (tmp == 1);
  //read_attribute(group, "redundant", tmp);
  //redundant_ = (tmp == 1);
  ...
```

then the code compiles and seems to run fine. However, I'm not sure what the function of the `redundant' attribute is, and wanted to check about how to proceed before putting in a PR to fix. I.e., if removing those lines is fine, or if some sort of test should be done on the object first before trying to load 'redundant' instead.

**Comment by jtramm** [comment]
A more robust fix might be to change the read_attr() function in hdf5_interface.cpp from:

```C++
void
read_attr(hid_t obj_id, const char* name, hid_t mem_type_id, void* buffer)
{
    hid_t attr = H5Aopen(obj_id, name, H5P_DEFAULT);
    H5Aread(attr, mem_type_id, buffer);
    H5Aclose(attr);
}
```

to:
```C++
void
read_attr(hid_t obj_id, const char* name, hid_t mem_type_id, void* buffer)
{
  htri_t exists = H5Aexists(obj_id, name);
  if( exists > 0 )
  {
    hid_t attr = H5Aopen(obj_id, name, H5P_DEFAULT);
    H5Aread(attr, mem_type_id, buffer);
    H5Aclose(attr);
  }
  else
    *((int *) buffer) = 0;
}
```

which seems to work for me.

**Comment by amandalund** [comment]
The "redundant" attribute was introduced in #1068 -- it's used to track which reactions are summed reactions for photon production. Have you recreated the HDF5 library since this PR was merged?

**Comment by paulromano** [comment]
Let me propose a simple solution -- add a check for the redundant attribute and if it doesn't exist, just set `redundant_` to false. This will allow the current version of OpenMC to work with older libraries (that don't store redundant reactions).

**Comment by jtramm** [comment]
@paulromano -- sounds good, it looks to me like that's what I'm doing in the read_attr() fix, but let me know if you had something else in mind or if I'm missing something.

**Comment by paulromano** [comment]
Yes, what you propose would work, but it might hide bugs too because then a 0-like value would be assigned to any missing attribute.

**Comment by jtramm** [comment]
Ah, yes, I was just looking at how it was being used for the redundant call (where the output is compared against 1), I didn't look at the rest of the use cases where the value is used more generally. To keep things so that it crashes if there is a missing attribute of another type besides "redundant", so it's obvious if there was an error generating/loading the library, I think this is what you had in mind:

```C++
Reaction::Reaction(hid_t group, const std::vector<int>& temperatures)
{
  read_attribute(group, "Q_value", q_value_);
  read_attribute(group, "mt", mt_);
  int tmp;
  read_attribute(group, "center_of_mass", tmp);
  scatter_in_cm_ = (tmp == 1);

  // Checks if redudant attribute exists before loading
  // (for compatibiltiy with legacy .h5 libraries)
  htri_t exists = H5Aexists(group, "redundant");
  if( exists )
  {
    read_attribute(group, "redundant", tmp);
    redundant_ = (tmp == 1);
  }
  else
    redundant_ = false;
```

Let me know if that works and I'll put in the PR

**Comment by jtramm** [comment]
Fixed with PR #1073.
